### PR TITLE
Treat circular reference when printing cells

### PR DESCRIPTION
**Pull request by ronisbr**
Test code with DataFrames:

```julia
julia> using DataFrames

julia> df = DataFrame(x = Any[missing])
1×1 DataFrame
 Row │ x       
     │ Any     
─────┼─────────
   1 │ missing 

julia> df[1, 1] = [df]
1-element Vector{DataFrame}:
 1×1 DataFrame
 Row │ x                                 
     │ Any                               
─────┼───────────────────────────────────
   1 │ DataFrame[#= circular reference …

julia> show(df, truncate = 100)
1×1 DataFrame
 Row │ x                                   
     │ Any                                 
─────┼─────────────────────────────────────
   1 │ DataFrame[#= circular reference =#]
```

### CODE REVIEW

**Comment by ronisbr** [comment]
Hi @bkamins!

This PR should fix the circular dependency in DataFrames. Can you take a brief look at the code to see if it was the best approach?

Thanks!

File: precompilation/precompile_PrettyTables.jl [hunk 1]
```diff
@@ -1,164 +1,162 @@
...
-    Base.precompile(Tuple{typeof(_process_data_cell_text),ProcessedTable,Int64,String,Int64,Int64,Int64,Int64,Crayon,Symbol,Ref{Any}})   # time: 0.003797233
-    Base.precompile(Tuple{Core.kwftype(typeof(_parse_cell_text)),NamedTuple{(:autowrap, :cell_data_type, :cell_first_line_only, :column_width, :compact_printing, :has_color, :limit_printing, :linebreaks, :renderer), Tuple{Bool, DataType, Bool, Int64, Bool, Bool, Bool, Bool, Val{:print}}},typeof(_parse_cell_text),UrlTextCell})   # time: 0.003454054
-    Base.precompile(Tuple{typeof(_print_custom_text_cell!),Display,UrlTextCell,String,Crayon,Int64,Ref{Any}})   # time: 0.001846952
-    Base.precompile(Tuple{Core.kwftype(typeof(_parse_cell_html)),NamedTuple{(:allow_html_in_cells, :cell_first_line_only, :compact_printing, :limit_printing, :linebreaks, :renderer), Tuple{Bool, Bool, Bool, Bool, Bool, Val{:print}}},typeof(_parse_cell_html),String})   # time: 0.001669777
-    Base.precompile(Tuple{typeof(_print_custom_text_cell!),Display,AnsiTextCell,String,Crayon,Int64,Ref{Any}})   # time: 0.001618891
+    Base.precompile(Tuple{typeof(_pt_text),IOContext{Base.TTY},PrintInfo})   # time: 0.22947502
```
Comment (bkamins) [inline precompilation/precompile_PrettyTables.jl:3]: You could check if using SnoopPrecompile.jl instead would not reduce the amount of code needed. See e.g. https://github.com/JuliaData/DataFrames.jl/blob/main/src/other/precompile.jl

(this is an independent thing from this PR)

File: src/backends/html/circular_reference.jl [hunk 2]
```diff
@@ -0,0 +1,12 @@
+# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
+#
+# Description
+# ==============================================================================
+#
+#   Functions to print to the IO that we have a circular reference.
+#
+# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
+
+function _pt_html_circular_reference(io::IOContext)
```
Comment (bkamins) [inline src/backends/html/circular_reference.jl:10]: not coverted by tests

File: src/backends/latex/circular_reference.jl [hunk 3]
```diff
@@ -0,0 +1,12 @@
+# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
+#
+# Description
+# ==============================================================================
+#
+#   Functions to print to the IO that we have a circular reference.
+#
+# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
+
+function _pt_latex_circular_reference(io::IOContext)
```
Comment (bkamins) [inline src/backends/latex/circular_reference.jl:10]: same here - add coverage

File: src/private.jl [hunk 4]
```diff
@@ -369,6 +370,36 @@ function _pt(
...
+        # PrettyTables.jl. Hence, we need to search if the current one is inside
+        # this vector. If true, we have a circular dependency.
+        for d in ptd
+            if d === _getdata(data)
+
+                if backend === Val(:text)
+                    _pt_text_circular_reference(context)
+                elseif backend === Val(:html)
+                    _pt_html_circular_reference(context)
+                elseif backend === Val(:latex)
```
Comment (bkamins) [inline src/private.jl:389]: below test coverage is missing in several places

File: src/private.jl [hunk 5]
```diff
@@ -369,6 +370,36 @@ function _pt(
...
+                    _pt_latex_circular_reference(context)
+                end
+
+                return nothing
+            end
+        end
+
+        # Otherwise, we must push the current data to the vector.
+        push!(ptd, _getdata(data))
```
Comment (bkamins) [inline src/private.jl:398]: especially testing this line is I think relevant.

File: test/general/circular_reference.jl [hunk 6]
```diff
@@ -0,0 +1,111 @@
+# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
+#
+# Description
+# ==============================================================================
+#
+#   Tests related with circular reference.
+#
+# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
```
Comment (bkamins) [inline test/general/circular_reference.jl:8]: I would add tests for:
```
m = [1 2 3]
m[1] = m;
pretty_table(m)
v = [1, 2, 3]
v[1] = v
pretty_table(v)
```

File: src/private.jl [hunk 7]
```diff
@@ -369,6 +370,36 @@ function _pt(
         end
     end
 
+    # Verify if we have a circular reference.
+    ptd = get(io, :__PRETTY_TABLES_DATA__, nothing)
```
Comment (bkamins) [inline src/private.jl:374]: Do I understand the logic correctly that this records all objects that were passed for printing by `pretty_table`? Then for `DataFrame` it is OK, but what if the default printing of some object does not use `pretty_table` (probably this cannot be helped and one of the tests I proposed below will error, but I want to be sure I understand the logic)?

Things look good in general. Thank you!

File: precompilation/precompile_PrettyTables.jl (hunk 1 above)
Comment (ronisbr) [reply precompilation/precompile_PrettyTables.jl:3]: Thanks! I did not know about SnoopPrecompile until now :D Only SnoopCompile

File: src/backends/html/circular_reference.jl (hunk 2 above)
Comment (ronisbr) [reply src/backends/html/circular_reference.jl:10]: Thanks!

File: src/backends/latex/circular_reference.jl (hunk 3 above)
Comment (ronisbr) [reply src/backends/latex/circular_reference.jl:10]: Thanks!

File: src/private.jl (hunk 7 above)
Comment (ronisbr) [reply src/private.jl:374]: Thanks!

My approach is:

Every time an object is printed with PrettyTables, I add it in this vector. All cells are rendered using an `IOContext` with this vector. Thus, any object that calls `pretty_table` is recorded.

If the default printing of an object does not use `pretty_table`, I do not know if we can do something about it.

Comment (bkamins) [reply src/private.jl:374]: > I do not know if we can do something about it.

We cannot. This is what I thought you do. So all is OK.

Comment (bkamins) [reply src/private.jl:374]: The reason why your code is OK is that immutable values cannot currently be self-refferential in Julia.

Comment (ronisbr) [reply src/private.jl:374]: Yes! I though about it and then I realize that it is just not a problem.

**Comment by ronisbr** [comment]
Thanks @bkamins ! I implemented all your suggestions and also found some bugs :D PrettyTables can now handle circular dependencies in DataFrames.

**Comment by bkamins** [comment]
Thank you!

**Pull request merged**